org.eclipse.lsp4e/src/org/eclipse/lsp4e/Canceller.java

```diff
@@ -0,0 +1,74 @@
+package org.eclipse.lsp4e;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.eclipse.lsp4j.services.LanguageServer;
+
+public class Canceller {
```

Comment:
I think there is indeed a need for such logic somewhere when we create the "compound" futures from the request to propagate cancellation. In my draft attempt locally, I try to have it as a method, but a dedicated class may be a better fit.
Couldn't it be made package visible only (ie not public)?

Reply:
Yes, I don't think there is any need for this to be public API.